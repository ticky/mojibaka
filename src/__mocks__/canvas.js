export const prepareCanvasContext = jest.fn((context) => context);
export const canDrawCharacter = jest.fn();
export const getCharacterWidth = jest.fn();
