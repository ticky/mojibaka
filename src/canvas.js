const DRAW_SIZE = 32;

export function prepareCanvasContext(context) {
  if (!context) {
    context = document.createElement('canvas').getContext('2d');
    context.fillStyle = "#000";
    context.font = `${DRAW_SIZE}px "Apple Color Emoji","Segoe UI","Segoe UI Emoji","Segoe UI Symbol",Arial`;
  } else {
    context.clearRect(0, 0, DRAW_SIZE * 2, DRAW_SIZE * 2);
  }
  return context;
}

export function canDrawCharacter(character, context) {
  context = prepareCanvasContext(context);
  context.fillText(character, 0, DRAW_SIZE);

  return context.getImageData(DRAW_SIZE / 2, DRAW_SIZE / 2, 1, 1).data[0] !== 0;
}

export function getCharacterWidth(character, context) {
  context = prepareCanvasContext(context);

  return context.measureText(character).width;
}
