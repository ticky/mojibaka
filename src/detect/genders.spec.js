/* global jest, describe, beforeEach, expect, it */
import detectGenders from './genders';
import { prepareCanvasContext, getCharacterWidth } from '../canvas';

jest.mock('../canvas');

describe('detectGenders', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('with an existing canvas context', () => {
    it('calls through to `prepareCanvasContext`', () => {
      const context = {};
      detectGenders(context);
      expect(prepareCanvasContext).toHaveBeenCalledTimes(1);
      expect(prepareCanvasContext).toHaveBeenCalledWith(context);
    });

    it('calls through to `getCharacterWidth`', () => {
      const context = {};
      detectGenders(context);
      expect(getCharacterWidth).toHaveBeenCalledTimes(2);
      expect(getCharacterWidth.mock.calls).toMatchSnapshot();
    });
  });

  describe('with no supplied canvas context', () => {
    it('calls through to `prepareCanvasContext`', () => {
      detectGenders();
      expect(prepareCanvasContext).toHaveBeenCalledTimes(1);
      expect(prepareCanvasContext).toHaveBeenCalledWith(undefined);
    });

    it('calls through to `getCharacterWidth`', () => {
      detectGenders();
      expect(getCharacterWidth).toHaveBeenCalledTimes(2);
      expect(getCharacterWidth.mock.calls).toMatchSnapshot();
    });
  });

  it('returns expected values on OS with differing fitzpatrick glyph sizes', () => {
    getCharacterWidth
      .mockImplementation((char) => {
        if (char === '🕵') {
          return 32;
        } else if (char === '🕵🏻‍♀️') {
          return 64;
        }
      });
    expect(detectGenders()).toBe(false);
  });

  it('returns expected values on OS with matching fitzpatrick glyph sizes, and with float glyph sizes', () => {
    getCharacterWidth
      .mockImplementation((char) => {
        if (char === '🕵') {
          return 32;
        } else if (char === '🕵🏻‍♀️') {
          return 32;
        }
      });
    expect(detectGenders()).toBe(true);

    getCharacterWidth
      .mockImplementation((char) => {
        if (char === '🕵') {
          return 43.939998626708984;
        } else if (char === '🕵🏻‍♀️') {
          return 43.939998626708984;
        }
      });
    expect(detectGenders()).toBe(true);
  });
});