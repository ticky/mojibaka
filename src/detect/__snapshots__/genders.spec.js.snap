exports[`detectGenders with an existing canvas context calls through to \`getCharacterWidth\` 1`] = `
Array [
  Array [
    "🕵",
    Object {},
  ],
  Array [
    "🕵🏻‍♀️",
    Object {},
  ],
]
`;

exports[`detectGenders with no supplied canvas context calls through to \`getCharacterWidth\` 1`] = `
Array [
  Array [
    "🕵",
    undefined,
  ],
  Array [
    "🕵🏻‍♀️",
    undefined,
  ],
]
`;
